File: services/match-service/internal/matchstate/reduce_test.go
package matchstate_test

import (
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"github.com/AndoNorth/match-flow/services/match-service/internal/matchstate"
)

var _ = Describe("Reduce", func() {
	baseState := matchstate.State{Status: "scheduled", LastSequence: 0}

	It("sets status for MatchStart", func() {
		rule := matchstate.Rule{Category: matchstate.MatchStart, Status: "live"}
		event := matchstate.Event{Sequence: 1}

		next, ok := matchstate.Reduce(baseState, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.Status).To(Equal("live"))
		Expect(next.LastSequence).To(Equal(1))
	})

	It("sets status for PeriodBoundary", func() {
		rule := matchstate.Rule{Category: matchstate.PeriodBoundary, Status: "half_time"}
		event := matchstate.Event{Sequence: 2}

		next, ok := matchstate.Reduce(baseState, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.Status).To(Equal("half_time"))
	})

	It("sets status for MatchEnd", func() {
		rule := matchstate.Rule{Category: matchstate.MatchEnd, Status: "full_time"}
		event := matchstate.Event{Sequence: 3}

		next, ok := matchstate.Reduce(baseState, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.Status).To(Equal("full_time"))
	})

	It("increments the home score and sets the clock for a ScoreEvent", func() {
		rule := matchstate.Rule{Category: matchstate.ScoreEvent}
		event := matchstate.Event{
			Sequence: 4,
			Payload:  map[string]any{"team": "home", "minute": 23},
		}

		next, ok := matchstate.Reduce(baseState, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.HomeScore).To(Equal(1))
		Expect(next.AwayScore).To(Equal(0))
		Expect(next.ClockMins).To(Equal(23))
	})

	It("increments the away score for a ScoreEvent", func() {
		rule := matchstate.Rule{Category: matchstate.ScoreEvent}
		event := matchstate.Event{
			Sequence: 5,
			Payload:  map[string]any{"team": "away", "minute": 40},
		}

		next, ok := matchstate.Reduce(baseState, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.AwayScore).To(Equal(1))
	})

	It("handles a JSON-decoded float64 minute the same as an int", func() {
		rule := matchstate.Rule{Category: matchstate.ScoreEvent}
		event := matchstate.Event{
			Sequence: 6,
			Payload:  map[string]any{"team": "home", "minute": float64(55)},
		}

		next, ok := matchstate.Reduce(baseState, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.ClockMins).To(Equal(55))
	})

	It("causes no state change for Unknown (e.g. card)", func() {
		rule := matchstate.Rule{Category: matchstate.Unknown}
		event := matchstate.Event{
			Sequence: 7,
			Payload:  map[string]any{"team": "home", "minute": 60},
		}
		state := matchstate.State{Status: "live", HomeScore: 1, ClockMins: 23, LastSequence: 6}

		next, ok := matchstate.Reduce(state, rule, event)

		Expect(ok).To(BeTrue())
		Expect(next.Status).To(Equal("live"))
		Expect(next.HomeScore).To(Equal(1))
		Expect(next.ClockMins).To(Equal(23)) // unchanged - card never touches clock_mins
		Expect(next.LastSequence).To(Equal(7))
	})

	It("skips an event whose Sequence is not greater than LastSequence", func() {
		rule := matchstate.Rule{Category: matchstate.ScoreEvent}
		event := matchstate.Event{Sequence: 3, Payload: map[string]any{"team": "home"}}
		state := matchstate.State{HomeScore: 2, LastSequence: 5}

		next, ok := matchstate.Reduce(state, rule, event)

		Expect(ok).To(BeFalse())
		Expect(next).To(Equal(state)) // unchanged
	})
})

File: frontend/lib/gateway/reduce.test.ts
import { describe, expect, it } from "vitest";
import { liveClockMins, reduce } from "./reduce";
import type { MatchBody } from "./types";

const base: MatchBody = {
  match_id: "m1",
  sport: "football",
  status: "scheduled",
  home_team: "Ashford United",
  away_team: "Denbury City",
  home_score: 0,
  away_score: 0,
  clock_mins: 0,
  created_at: "2026-08-31T12:00:00Z",
};

describe("reduce", () => {
  it("kickoff sets status to live, no score change", () => {
    const next = reduce(base, { type: "kickoff", sequence: 1, payload: {} });
    expect(next).toEqual({ ...base, status: "live" });
  });

  it("half_time sets status to half_time", () => {
    const next = reduce(
      { ...base, status: "live" },
      { type: "half_time", sequence: 5, payload: {} },
    );
    expect(next.status).toBe("half_time");
  });

  it("full_time sets status to full_time", () => {
    const next = reduce(
      { ...base, status: "live" },
      { type: "full_time", sequence: 10, payload: {} },
    );
    expect(next.status).toBe("full_time");
  });

  it("a home goal increments home_score and sets clock_mins from the payload", () => {
    const live = { ...base, status: "live" as const };
    const next = reduce(live, {
      type: "goal",
      sequence: 3,
      payload: { team: "home", minute: 23 },
    });
    expect(next).toEqual({ ...live, home_score: 1, clock_mins: 23 });
  });

  it("an away goal increments away_score", () => {
    const live = { ...base, status: "live" as const };
    const next = reduce(live, {
      type: "goal",
      sequence: 3,
      payload: { team: "away", minute: 40 },
    });
    expect(next.away_score).toBe(1);
    expect(next.home_score).toBe(0);
  });

  it("a card event changes nothing on the match state", () => {
    const live = { ...base, status: "live" as const };
    const next = reduce(live, {
      type: "card",
      sequence: 4,
      payload: { team: "home", minute: 30 },
    });
    expect(next).toEqual(live);
  });

  it("an unrecognized event type changes nothing", () => {
    const live = { ...base, status: "live" as const };
    const next = reduce(live, {
      type: "something_new",
      sequence: 4,
      payload: {},
    });
    expect(next).toEqual(live);
  });
});

describe("liveClockMins", () => {
  const kickoffAt = new Date(base.created_at).getTime();

  it("stays at 0 for a scheduled match regardless of elapsed time", () => {
    const scheduled = { ...base, status: "scheduled" as const };
    expect(liveClockMins(scheduled, kickoffAt + 60_000)).toBe(0);
  });

  it("infers the elapsed minute for a live match, one minute per second", () => {
    const live = { ...base, status: "live" as const, clock_mins: 0 };
    expect(liveClockMins(live, kickoffAt + 30_000)).toBe(30);
  });

  it("never infers below the last known authoritative clock_mins", () => {
    const live = { ...base, status: "live" as const, clock_mins: 50 };
    expect(liveClockMins(live, kickoffAt + 10_000)).toBe(50);
  });

  it("caps inference at 90 minutes for a still-live match", () => {
    const live = { ...base, status: "live" as const, clock_mins: 0 };
    expect(liveClockMins(live, kickoffAt + 200_000)).toBe(90);
  });

  it("freezes at clock_mins once full_time, ignoring elapsed time", () => {
    const finished = { ...base, status: "full_time" as const, clock_mins: 90 };
    expect(liveClockMins(finished, kickoffAt + 500_000)).toBe(90);
  });
});
